frontend/catalyst/device/verification.py

```diff
@@ -298,7 +307,25 @@ def _obs_checker(obs):
             )
 
     for m in tape.measurements:
+        # verify observable is supported
         if m.obs:
             _verify_observable(m.obs, _obs_checker)
+        # verify measurement process type is supported
+        mp_name = m.return_type.value if m.return_type else type(m).__name__
+        if not mp_name.title() in qjit_capabilities.measurement_processes:
+            if shots and not isinstance(m, SampleMeasurement):
+                raise CompileError(
+                    f"State-based measurements like {m} cannot work with finite shots. "
+                    "Please specify shots=None."
+                )
+            if not shots and not isinstance(m, StateMeasurement):
+                raise CompileError(
+                    f"Sample-based measurements like {m} cannot work with shots=None. "
+                    "Please specify a finite number of shots."
+                )
```

Comment:
Should these to checks be performed regardless of whether the MP is in the capabilities?

Reply:
That's an interesting point. In theory, that should always be redundant, because those cases should be removed from the capabilities. The case it _would_ catch is people making the TOML files incorrectly, for example by specifying that `Sample` is a valid measurement but failing to specify its conditions. 

Seems safer to just always check, I'll move them :)